File: TinkerRocketApp/TinkerRocketApp/Maps/Offline/TileSource.swift
//
//  TileSource.swift
//  TinkerRocketApp
//
//  Offline maps — see docs/plans/offline-maps.md.
//
//  Selectable map imagery sources, shared by the Rocket Map and (later) Drift
//  Cast. Apple cases use the built-in MapKit basemap, which CANNOT be cached for
//  offline use (no public API; against MapKit terms) — they stay as the global
//  online option. The USGS cases are public-domain tiles drawn via a
//  basemap-replacing MKTileOverlay and are the only sources we cache/download
//  for offline use. (Esri World Imagery looked great online but its free terms
//  restrict offline caching, which is the point of this feature — see the
//  international-imagery follow-up issue.)
//

import MapKit

enum TileSource: String, CaseIterable, Identifiable {
    case appleStandard
    case appleHybrid
    case usgsImagery
    case usgsImageryTopo
    case usgsTopo

    var id: String { rawValue }

    var displayName: String {
        switch self {
        case .appleStandard:   return "Apple Standard"
        case .appleHybrid:     return "Apple Hybrid"
        case .usgsImagery:     return "USGS Imagery"
        case .usgsImageryTopo: return "USGS Imagery + Topo"
        case .usgsTopo:        return "USGS Topo"
        }
    }

    /// SF Symbol for the source menu.
    var symbol: String {
        switch self {
        case .appleStandard:   return "map"
        case .appleHybrid:     return "globe.americas.fill"
        case .usgsImagery:     return "globe.desk.fill"
        case .usgsImageryTopo: return "mountain.2.fill"
        case .usgsTopo:        return "map.fill"
        }
    }

    /// Apple basemap cases render natively; USGS cases draw via MKTileOverlay.
    var isAppleBasemap: Bool { self == .appleStandard || self == .appleHybrid }

    /// Whether this source can be cached / downloaded for offline use.
    /// USGS (public domain) yes; Apple no.
    var isCacheable: Bool { !isAppleBasemap }

    /// Deepest zoom this source actually serves.  The USGS ArcGIS basemaps
    /// stop at 16 — asking for 17/18 returns 404 for every tile in the band,
    /// which the downloader counts as done with 0 bytes, so an over-deep
    /// request silently inflates the estimate and downloads nothing.
    /// Matches Android's per-source `TileSource.maxZoom`.
    var maxZoom: Int {
        isAppleBasemap ? 19 : 16
    }

    /// MKMapType for the Apple cases. For USGS cases the value is irrelevant
    /// (the overlay replaces the basemap) but `.standard` keeps Apple from
    /// fetching its own imagery underneath.
    var appleMapType: MKMapType {
        self == .appleHybrid ? .hybrid : .standard
    }

    /// `{x}`/`{y}`/`{z}` URL template for USGS cases, nil for the Apple basemap.
    /// ArcGIS REST tile endpoints are ordered `z/y/x`.
    var urlTemplate: String? {
        switch self {
        case .appleStandard, .appleHybrid:
            return nil
        case .usgsImagery:
            return "https://basemap.nationalmap.gov/arcgis/rest/services/USGSImageryOnly/MapServer/tile/{z}/{y}/{x}"
        case .usgsImageryTopo:
            return "https://basemap.nationalmap.gov/arcgis/rest/services/USGSImageryTopo/MapServer/tile/{z}/{y}/{x}"
        case .usgsTopo:
            return "https://basemap.nationalmap.gov/arcgis/rest/services/USGSTopo/MapServer/tile/{z}/{y}/{x}"
        }
    }

    /// Short attribution shown while the source is active (provider terms).
    var attribution: String? {
        isAppleBasemap ? nil : "USGS · The National Map"
    }
}

File: TinkerRocketAndroid/core/maps/src/main/kotlin/com/tinkerbug/tinkerrocket/maps/TileSource.kt
package com.tinkerbug.tinkerrocket.maps

/**
 * Selectable map imagery sources — port of iOS TileSource.swift.  The Apple
 * basemap cases have no Android equivalent and are omitted; USGS tiles are
 * public domain and are the only sources cached for offline use (the point
 * of the feature — see the iOS file for the Esri/Google rationale).
 *
 * GOOGLE_SATELLITE plays the iOS "Apple basemap" role on Android: a global,
 * online-only imagery layer whose provider terms forbid persistence.  It is
 * `cacheable = false` with a null template — the proxy fetches it through
 * [GoogleTileUpstream] (Map Tiles API session tokens) and never touches the
 * offline cache in either direction.
 *
 * The `key` doubles as the on-disk cache directory name and MUST match the
 * iOS rawValue so a future cache import/export stays byte-compatible.
 */
public enum class TileSource(
    public val key: String,
    public val displayName: String,
    /**
     * Upstream URL template.  ArcGIS REST tile endpoints are ordered
     * `z/y/x` — NOT the slippy-map `z/x/y` used on disk and by MapLibre
     * templates; the proxy owns that transposition.  Null for sources
     * fetched via a dedicated authenticated upstream instead of template
     * substitution — null also drops the source from every default
     * template map, so it can never be region-downloaded.
     */
    public val urlTemplate: String?,
    /** Provider's max native zoom; the map may overzoom beyond it. */
    public val maxZoom: Int,
    /**
     * Whether tiles may be persisted to the offline cache (mirrors iOS
     * `isCacheable`).  False = online-only by provider terms: the proxy
     * neither reads nor writes the cache for this source and stamps
     * responses `Cache-Control: no-store`.
     */
    public val cacheable: Boolean = true,
) {
    USGS_IMAGERY(
        key = "usgsImagery",
        displayName = "USGS Imagery",
        urlTemplate = "https://basemap.nationalmap.gov/arcgis/rest/services/USGSImageryOnly/MapServer/tile/{z}/{y}/{x}",
        maxZoom = 16,
    ),
    USGS_IMAGERY_TOPO(
        key = "usgsImageryTopo",
        displayName = "USGS Imagery + Topo",
        urlTemplate = "https://basemap.nationalmap.gov/arcgis/rest/services/USGSImageryTopo/MapServer/tile/{z}/{y}/{x}",
        maxZoom = 16,
    ),
    USGS_TOPO(
        key = "usgsTopo",
        displayName = "USGS Topo",
        urlTemplate = "https://basemap.nationalmap.gov/arcgis/rest/services/USGSTopo/MapServer/tile/{z}/{y}/{x}",
        maxZoom = 16,
    ),
    GOOGLE_SATELLITE(
        key = "googleSatellite",
        displayName = "Google Satellite",
        urlTemplate = null,
        maxZoom = 22,
        cacheable = false,
    ),
    ;

    public companion object {
        public fun fromKey(key: String): TileSource? = entries.firstOrNull { it.key == key }

        /** Attribution shown while a USGS source is active (provider terms). */
        public const val ATTRIBUTION: String = "USGS · The National Map"

        /**
         * Fallback attribution while Google Satellite is active; the proxy's
         * `/meta/googleSatellite/attribution` route serves the live copyright
         * string (Map Tiles API display requirement) once a session is up.
         */
        public const val GOOGLE_ATTRIBUTION: String = "© Google"
    }
}
